File: game.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cstdlib>
#include <ctime>
#include <limits>
using namespace std;

// --- Utility Functions ---
void clearInputBuffer() {
    cin.clear();
    cin.ignore(numeric_limits<streamsize>::max(), '\n');
}

int random(int min, int max) {
    return rand() % (max - min + 1) + min;
}

// --- Character Base Class ---
class Character {
protected:
    string name;
    int health;
    int attack;
    int defense;

public:
    Character(string n, int h, int a, int d)
        : name(n), health(h), attack(a), defense(d) {}

    virtual void displayStats() const {
        cout << "Name: " << name << " | HP: " << health
             << " | ATK: " << attack << " | DEF: " << defense << endl;
    }

    bool isAlive() const {
        return health > 0;
    }

    virtual int takeDamage(int dmg) {
        int damageTaken = dmg - defense;
        if (damageTaken < 0) damageTaken = 0;
        health -= damageTaken;
        if (health < 0) health = 0;
        return damageTaken;
    }

    virtual int attackTarget(Character& target) {
        int damage = random(attack / 2, attack);
        return target.takeDamage(damage);
    }

    string getName() const {
        return name;
    }

    int getHealth() const {
        return health;
    }

    void heal(int amount) {
        health += amount;
        if (health > 100) health = 100;
    }
};

// --- Player Class ---
class Player : public Character {
private:
    vector<string> inventory;

public:
    Player(string n)
        : Character(n, 100, 20, 5) {
        inventory.push_back("Potion");
    }

    void showInventory() {
        cout << "\nInventory: ";
        if (inventory.empty()) {
            cout << "Empty\n";
        } else {
            for (const auto& item : inventory) {
                cout << item << " ";
            }
            cout << endl;
        }
    }

    void useItem() {
        if (inventory.empty()) {
            cout << "No items to use!\n";
            return;
        }

        cout << "Using " << inventory.back() << "...\n";
        if (inventory.back() == "Potion") {
            heal(30);
            cout << "Healed 30 HP!\n";
        }
        inventory.pop_back();
    }

    void addItem(string item) {
        inventory.push_back(item);
    }
};

// --- Enemy Class ---
class Enemy : public Character {
public:
    Enemy(string n, int h, int a, int d)
        : Character(n, h, a, d) {}

    static Enemy generateRandom() {
        vector<string> names = {"Goblin", "Orc", "Dark Elf", "Skeleton"};
        string name = names[random(0, names.size() - 1)];
        int health = random(50, 100);
        int attack = random(10, 25);
        int defense = random(2, 10);
        return Enemy(name, health, attack, defense);
    }
};

// --- Game Functions ---
void battle(Player& player, Enemy& enemy) {
    cout << "\nBattle begins between " << player.getName() << " and " << enemy.getName() << "!\n";
    while (player.isAlive() && enemy.isAlive()) {
        cout << "\n-- " << player.getName() << "'s Turn --\n";
        player.displayStats();
        enemy.displayStats();

        cout << "\n1. Attack\n2. Use Item\nChoose action: ";
        int choice;
        cin >> choice;
        clearInputBuffer();

        if (choice == 1) {
            int dmg = player.attackTarget(enemy);
            cout << "You dealt " << dmg << " damage!\n";
        } else if (choice == 2) {
            player.useItem();
        } else {
            cout << "Invalid choice.\n";
        }

        if (!enemy.isAlive()) break;

        cout << "\n-- " << enemy.getName() << "'s Turn --\n";
        int dmg = enemy.attackTarget(player);
        cout << enemy.getName() << " dealt " << dmg << " damage!\n";
    }

    if (player.isAlive()) {
        cout << "\nYou defeated " << enemy.getName() << "!\n";
        if (random(1, 100) <= 40) {
            player.addItem("Potion");
            cout << "You found a Potion!\n";
        }
    } else {
        cout << "\nYou were defeated...\n";
    }
}

// --- Main Game ---
int main() {
    srand((unsigned)time(0));

    cout << "Welcome to the RPG Battle Simulator!\nEnter your character's name: ";
    string playerName;
    getline(cin, playerName);

    Player player(playerName);

    while (true) {
        cout << "\n-- Main Menu --\n";
        cout << "1. Fight an enemy\n";
        cout << "2. Show stats\n";
        cout << "3. Inventory\n";
        cout << "4. Exit\n";
        cout << "Choose an action: ";
        int choice;
        cin >> choice;
        clearInputBuffer();

        switch (choice) {
            case 1:
                if (!player.isAlive()) {
                    cout << "You're too injured to fight.\n";
                } else {
                    Enemy enemy = Enemy::generateRandom();
                    battle(player, enemy);
                }
                break;
            case 2:
                player.displayStats();
                break;
            case 3:
                player.showInventory();
                break;
            case 4:
                cout << "Thanks for playing!\n";
                return 0;
            default:
                cout << "Invalid option.\n";
        }
    }

    return 0;
}

File: game.py
import random
import time
import sys

# --- Helper functions ---
def delay_print(s):
    """Print with delay for dramatic effect."""
    for c in s:
        sys.stdout.write(c)
        sys.stdout.flush()
        time.sleep(0.02)
    print()

# --- Character Class ---
class Character:
    def __init__(self, name, health, attack_power, inventory=None):
        self.name = name
        self.health = health
        self.attack_power = attack_power
        self.inventory = inventory if inventory else []

    def is_alive(self):
        return self.health > 0

    def attack(self, target):
        damage = random.randint(1, self.attack_power)
        delay_print(f"{self.name} attacks {target.name} for {damage} damage!")
        target.health -= damage
        return damage

    def heal(self):
        if "potion" in self.inventory:
            self.health += 20
            self.inventory.remove("potion")
            delay_print(f"{self.name} drinks a potion and heals 20 HP!")
        else:
            delay_print(f"{self.name} has no potion!")

# --- Enemy Generator ---
def generate_enemy(level):
    names = ["Skeleton", "Dark Wolf", "Goblin", "Orc", "Shade"]
    name = random.choice(names)
    health = 30 + level * 10
    attack = 5 + level * 2
    return Character(name, health, attack)

# --- Game Locations ---
def forest_path(player):
    delay_print("You walk through the eerie Whispering Forest...")
    if random.random() < 0.7:
        enemy = generate_enemy(1)
        combat(player, enemy)
    else:
        delay_print("You found a health potion on the ground!")
        player.inventory.append("potion")

def cursed_cave(player):
    delay_print("The cave is dark and smells of death...")
    if random.random() < 0.5:
        enemy = generate_enemy(2)
        combat(player, enemy)
    else:
        delay_print("You discover a hidden treasure chest!")
        reward = random.choice(["potion", "ancient sword"])
        player.inventory.append(reward)
        delay_print(f"You obtained: {reward}!")

# --- Combat Loop ---
def combat(player, enemy):
    delay_print(f"A wild {enemy.name} appears!")
    while player.is_alive() and enemy.is_alive():
        delay_print(f"\n{player.name}'s HP: {player.health}")
        delay_print(f"{enemy.name}'s HP: {enemy.health}")
        action = input("Choose action: [A]ttack, [H]eal, [R]un: ").lower()
        if action == "a":
            player.attack(enemy)
            if enemy.is_alive():
                enemy.attack(player)
        elif action == "h":
            player.heal()
            if enemy.is_alive():
                enemy.attack(player)
        elif action == "r":
            if random.random() < 0.5:
                delay_print("You escaped successfully!")
                return
            else:
                delay_print("You failed to escape!")
                enemy.attack(player)
        else:
            delay_print("Invalid action.")

    if not player.is_alive():
        delay_print("You have fallen in battle...")
        exit()
    else:
        delay_print(f"You defeated the {enemy.name}!")

# --- Game Intro ---
def game_intro():
    delay_print("Welcome to the Realm of the Forgotten King.")
    name = input("Enter your hero's name: ")
    player = Character(name, 100, 15, ["potion"])
    delay_print(f"Brave {player.name}, your journey begins now!")
    return player

# --- Main Game Loop ---
def main():
    player = game_intro()

    locations = [forest_path, cursed_cave]
    while True:
        print("\nWhere will you go?")
        print("1. Whispering Forest")
        print("2. Cursed Cave")
        print("3. View Stats")
        print("4. Quit")
        choice = input("Choose: ")

        if choice == "1":
            forest_path(player)
        elif choice == "2":
            cursed_cave(player)
        elif choice == "3":
            delay_print(f"\nName: {player.name}")
            delay_print(f"Health: {player.health}")
            delay_print(f"Attack Power: {player.attack_power}")
            delay_print(f"Inventory: {player.inventory}")
        elif choice == "4":
            delay_print("Thanks for playing. Farewell!")
            break
        else:
            delay_print("Invalid option.")

if __name__ == "__main__":
    main()

